contracts/Minter.sol

```diff
@@ -0,0 +1,174 @@
+    require(newAdmin != address(0), 'SPSMinter: Admin cannot be address 0');
+
+    token = IMintable(newToken);
+    lastMintBlock = startBlock;
+    admin = newAdmin;
+    maxToPoolPerBlock = newMaxToPoolPerBlock;
+
+    require(token.decimals() == 18, "SPSMinter: Token doesn't have 18 decimals");
+
+    emit UpdateAdmin(address(0), newAdmin);
+  }
+
+  /**
+   * @notice Mint tokens to all pools, can be called by anyone
+   */
+  function mint() public {
+    require(totalMinted != cap, "SPSMinter: Cap reached");
+    require(block.number > lastMintBlock, "SPSMinter: Mint block not yet reached");
+
+
+    uint256 mintDifference;
+    unchecked {
+      mintDifference = block.number - lastMintBlock;
+    }
+
+    lastMintBlock = block.number;
+
+    for (uint256 i = 0; i < pools.length; i++){
+      uint256 amount = pools[i].amountPerBlock * mintDifference;
+
+      if(totalMinted + amount >= cap){
+        amount = cap - totalMinted;
+      }
+
+      token.mint(pools[i].receiver, amount);
+      totalMinted = totalMinted + amount;
+
+      emit Mint(pools[i].receiver, amount);
+    }
+  }
+
+  /**
+   * @notice Add new pool, can be called by admin
+   * @param newReceiver Address of the receiver
+   * @param newAmount Amount of tokens per block
+   */
+  function addPool(address newReceiver, uint256 newAmount) external onlyAdmin {
+    require(pools.length <= poolsCap, 'SPSMinter: Pools cap reached');
+    require(newAmount <= maxToPoolPerBlock, 'SPSMinter: Maximum amount per block reached');
+    pools.push(Pool(newReceiver, newAmount));
+    emit PoolAdded(newReceiver, newAmount);
+  }
+
+  /**
+   * @notice Update pool, can be called by admin
+   * @param index Index in the array of the pool
+   * @param newReceiver Address of the receiver
+   * @param newAmount Amount of tokens per block
+   */
+  function updatePool(uint256 index, address newReceiver, uint256 newAmount) external onlyAdmin {
+    require(newAmount <= maxToPoolPerBlock, 'SPSMinter: Maximum amount per block reached');
+    mint();
+    pools[index] = Pool(newReceiver, newAmount);
+    emit PoolUpdated(index, newReceiver, newAmount);
+  }
+
+  /**
+   * @notice Remove pool, can be called by admin
+   * @param index Index in the array of the pool
+   */
+  function removePool(uint256 index) external onlyAdmin {
+    mint();
+    emit PoolRemoved(index, pools[index].receiver, pools[index].amountPerBlock);
+
+    pools[index] = pools[pools.length - 1];
+    pools.pop();
+  }
+
+  /**
+   * @notice Update admin address
+   * @param newAdmin Address of the new admin
+   */
+  function updateAdmin(address newAdmin) external onlyAdmin {
+    emit UpdateAdmin(admin, newAdmin);
+    admin = newAdmin;
+  }
+
+  /**
+   * @notice Update maximum amount that can be created per block
+   * @param newMaxToPoolPerBlock maximum number of token that is allowed to be minted per block
+   */
+  function updateMaxPerBlock(uint256 newMaxToPoolPerBlock) external onlyAdmin {
```

Comment:
I'm wondering if a sanity check on the amount here is called for as well. Right now this could be set to ridiculous values, for example something greater than the cap. It might be worth adding a check that the amount is less than some sane value, for example a calculated percentage of the total cap (1%? 10%?). Do you have any thoughts @MattyIce?

Reply:
Looks like this check was not added?

Reply:
if we call `updateMaxPerBlock` with a value lower than the old max, there could exist pools with `amount > maxPerBlock` that still mint at `amount`. Perhaps this function should call `mint()` and then iterate thru pools setting `amount = maxPerBlock` when `amount > maxPerBlock`. Another idea from @mds1 is to create a function called `mintOnePool` that you could call inside the iteration proposed above, which would eliminate the need to call `mint()` (and therefore eliminate 2x iteration)

Reply:
yea so like @wildmolasses said, if we want `updateMaxPerBlock` to also update pools above the new max, you can refactor to something like the below (wrote this quickly so please review it before using!)

![image](https://user-images.githubusercontent.com/17163988/152616704-ffc4d8ab-9294-4d83-a2fc-14bd1cec1e5c.png)

![image](https://user-images.githubusercontent.com/17163988/152616909-152d6eaf-8320-4cde-8857-838ad8df7755.png)

Reply:
I think we can just hard-code the maxToPoolPerBlock number to make this whole thing a lot simpler. If it needs to change for some reason then the `admin` can just replace the minter contract with a new one. I think we can set it to `50 SPS` and that should cover us for everything we plan to do while preventing an error from minting too many tokens.

Reply:
Updated, hardcoded 50 SPS limit: https://github.com/steem-monsters/sps-minter/commit/59339e51acc0c556a946742644af7abb77638988 & https://github.com/steem-monsters/sps-minter/commit/f5bd1b02316bf20178b787db304b2ccad47aed61

Reply:
Looks reasonable @fbslo, one minor nit: the `maxToPoolPerBlock` variable can now be declared as `constant`

Reply:
Resolved: https://github.com/steem-monsters/sps-minter/commit/9f4569dc0ce8356c3a7c2d4c5b679cde38f2c44f